### DISCUSSION TITLE
Extension assets build tools

### DISCUSSION THREAD

**Discussion by vividviolet** [Shopify/shopify-cli-extensions#19] [General]
Here are the options we've discussed so far around what approach to take for building an extension for the new `shopify-cli-extensions`.

### 1. Inline esbuild

Use Go's esbuild package: https://pkg.go.dev/github.com/evanw/esbuild

**Pros**
- we can have single binary without any node dependencies
- allows us to control the build and reduce partner errors 
- significantly faster performance than running a separate node process

**Cons**
- hard to customize the build process, ex. how do we support additional plugins? Maybe we come up with good enough defaults and provide some config options to get around this
- ESBuild Go plugins are not well supported compared to Node plugins. This might mean us having to write our own Go plugins for different file types. In the original prototype we tested adding GraphQL support for an extension and couldn't find a Go plugin for this use case. In the future if we decide to support other file types like `.vue`, we'd need to write our own Go plugin.

### 2. Esbuild as a separate node process

Go will invoke `yarn/npm build` separately to build assets. The build command is defined in `package.json` and by default runs `node scripts/build.js`. Both the command and the build script can be modified.

**Pros**
- developers can fully customize the build and swap out esbuild completely with whatever they want to use
- developers can add in node plugins as needed. If they are fine with using esbuild they can use the default script and only need to modify the `plugins` option
- allows us to have a very simple build.js script as a starting point without supporting any additional plugins

**Cons**
- requires pulling in the `esbuild` node module
- requires exposing a build.js script which could be daunting for users who don't need to know the details

### 3. Inline build + option to override the default build

Inline esbuild is used by default. We can also add a config option for overriding the inline build with a custom command

Ex. 
```
{
   build: 'webpack'
}
```

**Pros**
- hides complexity from the developers who just want to use the inline option
- allows full customization for power users

**Cons**
- added complexity with 2 code paths to maintain. The Go server needs to work with the inline build and also allow using a custom node build process

## Build tools options

**Esbuild**
- fast and compiles react/ts without any additional dependencies
- allows us to optionally inline the build process

**Webpack**
- This is what we're doing today. It's both slow, too customizable and needs to pull in too many dependencies

**Snowpack**
- Builds ESM modules, which are [not yet fully supported in Web Workers](https://www.snowpack.dev/guides/web-worker/#esm-web-worker-support) for all browsers. It might not be supported inside iOS JavaScriptCore either - have to check with the Mobile folks.

**SWC**
I'm not sure about this one - it looks like a transpiler and needs to use an additional bundler to work? Can you add more details @TheMallen ?

@TheMallen @lemonmade please add your thoughts and ideas. 

cc @cecyc @t6d @Fionoble @alanthai @olavoasantos @elanalynn

**Comment by olavoasantos** [comment]
As something in between option 1 and 2, we could try to execute JS plugins on esbuild GO ([this issue might help](https://github.com/evanw/esbuild/issues/821#issuecomment-797784638)). This way, if a user needs something beyond our defaults, they can install it and make use. Plus, it'd allow our Go code to make use of the JS plugins which has a larger community.

The biggest con here is that this is a custom bridge between the two that we'd have to write, which could lead unexpected bugs/challenges.

**Comment by olavoasantos** [comment]
As an alternative to 2, we could also bundle the JS version of esbuild into an executable using [something like pkg](https://github.com/vercel/pkg#readme). This is similar to what Go does, but it bundles the Node.js code into a single binary which can be executed without installing dependencies. Not sure if it's the best alternative, but I think it's worth mentioning.

**Comment by vividviolet** [comment]
In the interest of keeping the [current project](https://vault.shopify.io/projects/20476) in scope, we will implement a modified version of Option 2, I'll call it Option 4. Shipping Option 4 already achieves the majority of our project objectives:
- reduce external dependencies (babel and its plugins)
- remove `admin-ui-extensions-run` and `checkout-ui-extensions`

### 4. Esbuild as a separate node process with an inline build script

Go will invoke node and pipe the content of an internal build script to run esbuild. This still requires the developer to have `esbuild` in their node_modules. However, the build details are abstracted.

**Pros**
- allows us to control the build and reduce partner errors
- hides build details from the developers
- reduce branching in the code. Go will not need to run `npm` or `yarn`, it will simply call `node`

**Cons**
- requires pulling in the esbuild node module
- hard to customize the build process, ex. how do we support additional plugins?

Most of the work for this approach has already been done/prototyped. We can continue discussing options and iterate on the ideal solution later.

**Comment by lemonmade** [comment]
I didn't realize the Go version of esbuild is so limited in the available plugins. I think that changes what I think the best approach here is. First though, I wanted to state what I view as the main priorities, and things to avoid, in designing this new build system:

- We have to be able to optimize for performance (priority is making sure we can build the extension assets to be as small as possible, secondary is making sure that process is as fast as possible for the developer)
- We have to be able to extend our builds to provide default support for features that the community expects over time — the most important things I can think of are supporting environment variables, and being able to support other JavaScript frameworks if we build UI extension bindings for them (Vue being the obvious next one we could implement)
- We have to minimize the amount of dependencies and files in the developer's project that they would not know about or edit

Additionally, I think it's very important that, if `node`/ `yarn` are ever run, they are run by commands the developer can easily edit. We are going through a big transition in Node land to ESM being the default module system, but while we are on the way there, it is very hard to know how to run Node correctly on a project, especially if you include a source-controlled JavaScript file anywhere in the solution (since that JS file may be interpreted as CJS or ESM, depending on how the developer has configured their project).

Taking all those together, the approach I would recommend is this:

- We keep an `@shopify` dependency that is just responsible for builds; this will be installed as a development dependency for all template projects
- In the configuration file for the extension, there is a new `build: {command: 'my-build-command'}` option. If that command does not exist, we don’t build the project; the CLI does not offer a default. The template repo will have `yarn build` or `npm build` (ideally we'd choose the one we expect to be used), and would define the `build` script in `package.json` to call an executable provided by our dev dependency. Developers can easily change this command 
- Same general idea for serving, except it would be a separate top-level option in the config file, and would default to mapping to `yarn/npm start`
- That underlying build tool can be whatever we want. If the node version of `esbuild` is sufficient, we can use that. If not, I strongly suggest we use Rollup, which is our more modern choice for building library code at Shopify and is basically "webpack, the good parts". This is also closely aligned with our build process for Hydrogen, which uses Vite, which is just ESBuild+Rollup under the hood.

**Comment by vividviolet** [comment]
Given the limitations of the Go version and the ever-changing landscape of JS build tools, I agree with @lemonmade's proposal to give us the most flexibility and also future-proof our approach in case we change our build tool in the future.

We will be going with this approach:
- have a node package to build the extension
- the package will be responsible for pulling in `esbuild` and appropriate plugins to support the file types we care about (graphql, vue (future))
- have 2 scripts defined in the package.json for `serve` and `build` which the Go server will call using `yarn/npm`
- all build configurations (outDir, custom environment variables) will be set through a central YAML file instead of passed in as command line flags. We will also give developers a utility to grab the extension configs as JSON so they can use the configs with other build tools if needed

In the future (out of scope for the current project):
- allow developers to customize the build/serve command in the extension configs in case they don't want to use `yarn/npm`. 

Default:
```yml
build:
  command: 'yarn build'
```

Customized:
```yml
build:
  command: 'node ./scripts/build.js'
```
